File: java/src/test/java/com/google/cloud/crypto/tink/signature/PublicKeyVerifyFactoryTest.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.cloud.crypto.tink.signature;

import static junit.framework.Assert.fail;

import com.google.cloud.crypto.tink.CommonProto.EllipticCurveType;
import com.google.cloud.crypto.tink.CommonProto.HashType;
import com.google.cloud.crypto.tink.EcdsaProto.EcdsaPrivateKey;
import com.google.cloud.crypto.tink.EcdsaProto.EcdsaPublicKey;
import com.google.cloud.crypto.tink.EcdsaProto.EcdsaSignatureEncoding;
import com.google.cloud.crypto.tink.KeysetHandle;
import com.google.cloud.crypto.tink.PublicKeySign;
import com.google.cloud.crypto.tink.PublicKeyVerify;
import com.google.cloud.crypto.tink.TestUtil;
import com.google.cloud.crypto.tink.TinkProto.KeyData;
import com.google.cloud.crypto.tink.TinkProto.KeyStatusType;
import com.google.cloud.crypto.tink.TinkProto.Keyset.Key;
import com.google.cloud.crypto.tink.TinkProto.OutputPrefixType;
import com.google.cloud.crypto.tink.subtle.Random;
import java.security.GeneralSecurityException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Unit tests for {@link PublicKeyVerifyFactory}. */
@RunWith(JUnit4.class)
//TODO(quannguyen): Add more tests.
public class PublicKeyVerifyFactoryTest {

  @Before
  public void setUp() throws Exception {
    PublicKeyVerifyFactory.registerStandardKeyTypes();
    PublicKeySignFactory.registerStandardKeyTypes();
  }

  @Test
  public void testMultipleKeys() throws Exception {
    // Permutations of {0, 1, 2}.
    int[][] ids = new int[][] {
      {0, 1, 2},
      {0, 2, 1},
      {1, 0, 2},
      {1, 2, 0},
      {2, 0, 1},
      {2, 1, 0}
    };
    EcdsaPrivateKey[] ecdsaPrivKeys = new EcdsaPrivateKey[] {
      TestUtil.generateEcdsaPrivKey(EllipticCurveType.NIST_P521, HashType.SHA512,
          EcdsaSignatureEncoding.DER),
      TestUtil.generateEcdsaPrivKey(EllipticCurveType.NIST_P384, HashType.SHA512,
          EcdsaSignatureEncoding.DER),
      TestUtil.generateEcdsaPrivKey(EllipticCurveType.NIST_P256, HashType.SHA256,
          EcdsaSignatureEncoding.DER)};
    EcdsaPublicKey[] ecdsaPubKeys = new EcdsaPublicKey[] {
      ecdsaPrivKeys[0].getPublicKey(),
          ecdsaPrivKeys[1].getPublicKey(),
          ecdsaPrivKeys[2].getPublicKey()
    };
    Key[] keys = new Key[] {
      TestUtil.createKey(
          TestUtil.createKeyData(
              ecdsaPubKeys[0],
              "type.googleapis.com/google.cloud.crypto.tink.EcdsaPublicKey",
              KeyData.KeyMaterialType.ASYMMETRIC_PUBLIC),
          1,
          KeyStatusType.ENABLED,
          OutputPrefixType.TINK),
      TestUtil.createKey(
          TestUtil.createKeyData(
              ecdsaPubKeys[1],
              "type.googleapis.com/google.cloud.crypto.tink.EcdsaPublicKey",
              KeyData.KeyMaterialType.ASYMMETRIC_PUBLIC),
          2,
          KeyStatusType.ENABLED,
          OutputPrefixType.RAW),
      TestUtil.createKey(
          TestUtil.createKeyData(
              ecdsaPubKeys[2],
              "type.googleapis.com/google.cloud.crypto.tink.EcdsaPublicKey",
              KeyData.KeyMaterialType.ASYMMETRIC_PUBLIC),
          3,
          KeyStatusType.ENABLED,
          OutputPrefixType.LEGACY)};
    for (int i = 0; i < ids.length; i++) {
      KeysetHandle keysetHandle = TestUtil.createKeysetHandle(
        TestUtil.createKeyset(keys[ids[i][0]], keys[ids[i][1]], keys[ids[i][2]]));
      PublicKeyVerify verifier = PublicKeyVerifyFactory.getPrimitive(keysetHandle);
      // Signature signed by any of 3 keys should be verified by the verifier.
      for (int j = 0; j < 3; j++) {
        PublicKeySign signer = PublicKeySignFactory.getPrimitive(
          TestUtil.createKeysetHandle(
              TestUtil.createKeyset(
                  TestUtil.createKey(
                      TestUtil.createKeyData(
                          ecdsaPrivKeys[j],
                          "type.googleapis.com/google.cloud.crypto.tink.EcdsaPrivateKey",
                          KeyData.KeyMaterialType.ASYMMETRIC_PRIVATE),
                      j + 1,
                      KeyStatusType.ENABLED,
                      keys[j].getOutputPrefixType()))));
        byte[] plaintext = Random.randBytes(1211);
        byte[] sig = signer.sign(plaintext);
        try {
          verifier.verify(sig, plaintext);
        } catch (GeneralSecurityException ex) {
          fail("Valid signature, should not throw exception");
        }
      }
      // Signature signed by random key should not be verified by the verifier.
      EcdsaPrivateKey randomPrivKey = TestUtil.generateEcdsaPrivKey(
          EllipticCurveType.NIST_P521, HashType.SHA512, EcdsaSignatureEncoding.DER);
      PublicKeySign signer = PublicKeySignFactory.getPrimitive(
          TestUtil.createKeysetHandle(
              TestUtil.createKeyset(
                  TestUtil.createKey(
                      TestUtil.createKeyData(
                          randomPrivKey,
                          "type.googleapis.com/google.cloud.crypto.tink.EcdsaPrivateKey",
                          KeyData.KeyMaterialType.ASYMMETRIC_PRIVATE),
                      1,
                      KeyStatusType.ENABLED,
                      keys[0].getOutputPrefixType()))));
      byte[] plaintext = Random.randBytes(1211);
      byte[] sig = signer.sign(plaintext);
      try {
        verifier.verify(sig, plaintext);
        fail("Invalid signature, should have thrown exception");
      } catch (GeneralSecurityException expected) {
        // Expected
      }
    }
  }
}

File: cc/signature/public_key_verify_factory_test.cc
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/signature/public_key_verify_factory.h"

#include <string>
#include <utility>

#include "gtest/gtest.h"
#include "tink/config.h"
#include "tink/crypto_format.h"
#include "tink/keyset_handle.h"
#include "tink/public_key_verify.h"
#include "tink/registry.h"
#include "tink/signature/ecdsa_verify_key_manager.h"
#include "tink/signature/signature_config.h"
#include "tink/util/status.h"
#include "tink/util/test_keyset_handle.h"
#include "tink/util/test_util.h"
#include "proto/ecdsa.pb.h"
#include "proto/tink.pb.h"

using crypto::tink::TestKeysetHandle;
using crypto::tink::test::AddTinkKey;
using google::crypto::tink::EcdsaPublicKey;
using google::crypto::tink::EcdsaSignatureEncoding;
using google::crypto::tink::EllipticCurveType;
using google::crypto::tink::HashType;
using google::crypto::tink::KeyData;
using google::crypto::tink::Keyset;
using google::crypto::tink::KeyStatusType;

namespace crypto {
namespace tink {
namespace {

class PublicKeyVerifyFactoryTest : public ::testing::Test {
 protected:
  void SetUp() override {
    auto status = SignatureConfig::Register();
    ASSERT_TRUE(status.ok()) << status;
  }
};

EcdsaPublicKey GetNewEcdsaPublicKey() {
  auto ecdsa_key = test::GetEcdsaTestPrivateKey(EllipticCurveType::NIST_P256,
                                                HashType::SHA256,
                                                EcdsaSignatureEncoding::DER);
  return ecdsa_key.public_key();
}

TEST_F(PublicKeyVerifyFactoryTest, testBasic) {
  Keyset keyset;
  auto public_key_verify_result = PublicKeyVerifyFactory::GetPrimitive(
      *TestKeysetHandle::GetKeysetHandle(keyset));
  EXPECT_FALSE(public_key_verify_result.ok());
  EXPECT_EQ(absl::StatusCode::kInvalidArgument,
      public_key_verify_result.status().code());
  EXPECT_PRED_FORMAT2(testing::IsSubstring, "at least one key",
                      std::string(public_key_verify_result.status().message()));
}

TEST_F(PublicKeyVerifyFactoryTest, testPrimitive) {
  // Prepare a Keyset.
  Keyset keyset;
  std::string key_type =
      "type.googleapis.com/google.crypto.tink.EcdsaPublicKey";

  uint32_t key_id_1 = 1234543;
  AddTinkKey(key_type, key_id_1, GetNewEcdsaPublicKey(),
             KeyStatusType::ENABLED, KeyData::ASYMMETRIC_PUBLIC, &keyset);

  uint32_t key_id_2 = 726329;
  AddTinkKey(key_type, key_id_2, GetNewEcdsaPublicKey(),
             KeyStatusType::ENABLED, KeyData::ASYMMETRIC_PUBLIC, &keyset);

  uint32_t key_id_3 = 7213743;
  AddTinkKey(key_type, key_id_3, GetNewEcdsaPublicKey(),
             KeyStatusType::ENABLED, KeyData::ASYMMETRIC_PUBLIC, &keyset);

  keyset.set_primary_key_id(key_id_3);

  // Create a KeysetHandle and use it with the factory.
  auto public_key_verify_result = PublicKeyVerifyFactory::GetPrimitive(
      *TestKeysetHandle::GetKeysetHandle(keyset));
  EXPECT_TRUE(public_key_verify_result.ok())
      << public_key_verify_result.status();
  auto public_key_verify = std::move(public_key_verify_result.ValueOrDie());
}

}  // namespace
}  // namespace tink
}  // namespace crypto
